src/main/java/htsjdk/samtools/reference/BlockCompressedIndexedFastaSequenceFile.java

```diff
@@ -0,0 +1,103 @@
+ * SOFTWARE.
+ */
+
+package htsjdk.samtools.reference;
+
+import htsjdk.samtools.SAMException;
+import htsjdk.samtools.seekablestream.SeekablePathStream;
+import htsjdk.samtools.util.BlockCompressedInputStream;
+import htsjdk.samtools.util.GZIIndex;
+
+import java.io.BufferedInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+/**
+ * A block-compressed FASTA file driven by an index for fast, concurrent lookups.
+ *
+ * <p>Supports two interfaces: the ReferenceSequenceFile for old-style, stateful lookups and a direct getter.
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class BlockCompressedIndexedFastaSequenceFile extends AbstractIndexedFastaSequenceFile {
+
+    private final BlockCompressedInputStream stream;
+    private final GZIIndex gzindex;
+
+    public BlockCompressedIndexedFastaSequenceFile(Path path)
+            throws FileNotFoundException {
+        this(path,new FastaSequenceIndex((findRequiredFastaIndexFile(path))));
+    }
+
+    public BlockCompressedIndexedFastaSequenceFile(final Path path, final FastaSequenceIndex index) {
+        super(path, index);
+        if (!canCreateBlockCompresedIndexedFastaSequence(path)) {
+            throw new SAMException("Invalid block-compressed Fasta file");
+        }
+        try {
+            stream = new BlockCompressedInputStream(new SeekablePathStream(path));
+            gzindex = GZIIndex.loadIndex(GZIIndex.resolveIndexNameForBgzipFile(path));
+        } catch (IOException e) {
+            throw new SAMException("Fasta file should be readable but is not: " + path, e);
+        }
+    }
+
+    private boolean canCreateBlockCompresedIndexedFastaSequence(final Path path) {
+        try (final InputStream stream = new BufferedInputStream(Files.newInputStream(path))) {
+            // check if the it is a valid block-compressed file
+            if(BlockCompressedInputStream.isValidFile(stream)) {
+                // check if the .gzi index exits
+                return Files.exists(GZIIndex.resolveIndexNameForBgzipFile(path));
+            }
+        } catch (IOException e) {
+            return false;
+        }
+        return false;
+    }
+
+    @Override
+    protected int readFromPosition(final ByteBuffer buffer, final long position) throws IOException {
+        // old position to get back
+        final long oldPos = stream.getFilePointer();
+        try {
+            final long virtualOffset = gzindex.getVirtualOffsetForSeek(position);
+            stream.seek(virtualOffset);
+            final byte[] array = new byte[buffer.remaining()];
+            final int read = stream.read(array);
+            buffer.put(array);
+            return read;
+        } finally {
+            stream.seek(oldPos);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        stream.close();
```

Comment:
Compilation fails if so: "Abstract method close cannot be accessed directly" - this happens because the super in this case is the interface and not the abstract-class. Should I override for the abstract class with a no-op implementation?

Reply:
Oh, right, that makes sense.  Lets just leave it how it is.